# Descriptive Statistics Review

## Before you start:

- Read the README.md file
- Comment as much as you can
- Happy learning!

## Context

![img](./diamonds.jpg)

In this lab we are going to work with data to understand the characteristics of a diamond that are most likely to influence its price. In this first part of the lab, we will explore and clean our data. 

The dataset we will be using is comprised of approximately 54k rows and 11 different columns. As always, a row represents a single observation (in this case a diamond) and each of the columns represent a different feature of a diamond.

The following codebook was provided together with the dataset to clarify what each column represents:


| Column  | Description  |
|---|---|
| Price  | Price in US dollars (326-18,823)  |
| Carat  | Weight of the diamond (0.2--5.01)  |
| Cut  | Quality of the cut (Fair, Good, Very Good, Premium, Ideal)  |
| Color  | Diamond colour, from J (worst) to D (best)  |
| Clarity  | A measurement of how clear the diamond is (I1 (worst), SI2, SI1, VS2, VS1, VVS2, VVS1, IF (best))   |
| x  | Length in mm (0--10.74)  |
| y  | Width in mm (0--58.9)  |
| z  | Depth in mm (0--31.8)  |
| Depth  | Total depth percentage = z / mean(x, y) = 2 * z / (x + y) (43--79)  |
| Table  | Width of top of diamond relative to widest point (43--95)  |

## Libraries
Pandas and numpy will be needed for the analysis of the data. Don't worry about the seaborn and matplotlib import at the moment, you will learn more about them next week, but we will be using some of their functionalities.

In [101]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

First import the data from the .csv file provided and assign it to a variable named `diamonds` and drop the column with the index.

In [102]:
diamonds = pd.read_csv('/Users/bert/Desktop/dataV2-labs/module-2/M2-mini-project1/diamonds.csv').drop(['Unnamed: 0'], axis=1)

#.drop(['Unnamed: 0'], axis=1) --> this remove the column with the index. 
diamonds.head(10)

carat        cut color clarity  depth  table  price     x     y     z
0   0.23      Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1   0.21    Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
2   0.23       Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31
3   0.29    Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
4   0.31       Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75
5   0.24  Very Good     J    VVS2   62.8   57.0    336  3.94  3.96  2.48
6   0.24  Very Good     I    VVS1   62.3   57.0    336  3.95  3.98  2.47
7   0.26  Very Good     H     SI1   61.9   55.0    337  4.07  4.11  2.53
8   0.22       Fair     E     VS2   65.1   61.0    337  3.87  3.78  2.49
9   0.23  Very Good     H     VS1   59.4   61.0    338  4.00  4.05  2.39

# 1. Taking the first look at the data.
Let's see how the data looks by using pandas methods like `head()`, `info()` and `describe()`. 

**First, use the `head` method.**

In [103]:
diamonds.head()

carat      cut color clarity  depth  table  price     x     y     z
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
2   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31
3   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
4   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75

We can see the first 5 rows of the dataset using the `head` method. This by itself doesn't tell us much about the data that we have, but we can have a first look at the features (columns) and some of the values that each one takes.

**What do you see? Make some comments about the values you see in each column, comparing them with the codebook. Is that what you would expect for these variables?**

In [ ]:
'''
* We can see for each diamond the carat of the stone (= weight of the diamond) with min. (0.2 & max 5.01) --> here we have small diamonds
* Here we can also see how's the cut of the diamond is? (5 index columns where Ideal is the best and Premium second best. and good is just average.) If we only look to these 5 we have pretty good quality of diamonds. 
* The color is also important here we have 2 not so good qualities and 3 very good qualties just under the best quality. 
    (J --> D) --> D is the best quality and J is the worst!
* Clarity of these 5 stones is not so good if take a look to the codebook. We can say that those stones are average.  
* Depth: 
* Table: 
* Price: price of a diamond depends on a combination of several things. (Carat, Cut, Color, Clarity, Depth and Table)
* X: This represents the length of the diamond in mm of these 5 the range is between 3.89 and 4.34
* Y: This represenets the width of the diamond in mm of these 5 the range is between 3.84 and 4.45
* Z: This represents the depth of the diamond in mm of these 5 the range is between 2.31 and 2.75 

Was I expect thse variables? 
I would expect there was a larger difference in the price range. 
I was also expecting to see a relation between the values  Y in comparison with Table but at first glance we don't see much. 
I also thought that on the basis of 5 records, it was impossible to draw a conclusion. 

'''

It is very important to know the amount of data we have, because everything will depend on that, from the quality of the analysis to the choice of our infrastracture.

**Check the shape of the data**

In [104]:
diamonds.shape

(53940, 10)

The `clarity` column is confusing because we are not diamond experts. Let's create a new column with a new scale that is more understandable for us.

**Create a new column with numbers from 0 to 7. The lowest would be 0 with value `I1` and the greatest 7 with value `IF`**

In [105]:
def clarity_label (row): 
    if row['clarity'] == "I1": 
        return 0 
    if row['clarity'] == "SI2":
        return 1
    if row['clarity'] == "SI1":
        return 2
    if row['clarity'] == "VS2":
        return 3
    if row['clarity'] == "VS1":
        return 4
    if row['clarity'] == "VVS2":
        return 5
    if row['clarity'] == "VVS1":
        return 6
    if row['clarity'] == "IF":
        return 7

diamonds['clarity_scale'] = diamonds.apply(lambda row: clarity_label(row), axis=1)
diamonds.head()

#Comments: First I've made a function to specify every new variable. After I completed my function I used the 'Apply' method to the DataFrame and I used a lambda to iterate over all the rows. --> axis=1 means here in the apply function --> Apply function to each row. 

carat      cut color clarity  depth  table  price     x     y     z  \
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43   
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31   
2   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31   
3   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63   
4   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75   

   clarity_scale  
0              1  
1              2  
2              4  
3              3  
4              1

It makes sense to do the same with the `color` column.

**Do the same with values from 0 to 6. Read the codebook to see the match**

In [106]:
def color_label (row): 
    if row['color'] == "J": 
        return 0 
    if row['color'] == "I":
        return 1
    if row['color'] == "H":
        return 2
    if row['color'] == "G":
        return 3
    if row['color'] == "F":
        return 4
    if row['color'] == "E":
        return 5
    if row['color'] == "D":
        return 6

diamonds['color_scale'] = diamonds.apply(lambda row: color_label(row), axis=1)
diamonds.head()


carat      cut color clarity  depth  table  price     x     y     z  \
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43   
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31   
2   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31   
3   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63   
4   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75   

   clarity_scale  color_scale  
0              1            5  
1              2            5  
2              4            5  
3              3            1  
4              1            0

With the `info` method, we can see the features of the dataset, and the amount of observations (rows) that have a non-null value and the types of the features. 

**Now use the `info` method and comparing with the shape, comment on what you see**

In [108]:
diamonds.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53940 entries, 0 to 53939
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   carat          53940 non-null  float64
 1   cut            53940 non-null  object 
 2   color          53940 non-null  object 
 3   clarity        53940 non-null  object 
 4   depth          53940 non-null  float64
 5   table          53940 non-null  float64
 6   price          53940 non-null  int64  
 7   x              53940 non-null  float64
 8   y              53940 non-null  float64
 9   z              53940 non-null  float64
 10  clarity_scale  53940 non-null  int64  
 11  color_scale    53940 non-null  int64  
dtypes: float64(6), int64(3), object(3)
memory usage: 4.9+ MB


In [109]:
diamonds.shape

(53940, 12)

In [ ]:
"""
If we compare the info() method to shape() method we can see that number of rows identical and that is also what we expect. 
"""

In the last line of the info output, you have some information about the types of the columns. As you know, it is a good idea to check if the types of each column is what you expect. If a column has the right type, we will be able to do all the operations that we want to do. 

For instance, if we have a column that is a `date` with a `string` format, we will have the data but we won't be able to do a simple operation, such as format the date the way that we would like.

Changing the data type to the one we needs can help us to solve a lot of problems in our data.

**Check the types of each column and comment if it matches with the expected**

In [110]:
#Here I change the datatype for a price because a price is not always an integer. 
diamonds.astype({'price': 'float64'}).dtypes

carat            float64
cut               object
color             object
clarity           object
depth            float64
table            float64
price            float64
x                float64
y                float64
z                float64
clarity_scale      int64
color_scale        int64
dtype: object

# 2. A deeper look: checking the basic statistics.

The `describe` method gives us an overview of our data. From here we can see all the descriptive metrics for our variables.

**Use the `describe` method and comment on what you see**

In [111]:
diamonds.describe().transpose()

count         mean          std    min     25%      50%  \
carat          53940.0     0.797940     0.474011    0.2    0.40     0.70   
depth          53940.0    61.749405     1.432621   43.0   61.00    61.80   
table          53940.0    57.457184     2.234491   43.0   56.00    57.00   
price          53940.0  3932.799722  3989.439738  326.0  950.00  2401.00   
x              53940.0     5.731157     1.121761    0.0    4.71     5.70   
y              53940.0     5.734526     1.142135    0.0    4.72     5.71   
z              53940.0     3.538734     0.705699    0.0    2.91     3.53   
clarity_scale  53940.0     3.051020     1.647136    0.0    2.00     3.00   
color_scale    53940.0     3.405803     1.701105    0.0    2.00     3.00   

                   75%       max  
carat             1.04      5.01  
depth            62.50     79.00  
table            59.00     95.00  
price          5324.25  18823.00  
x                 6.54     10.74  
y                 6.54     58.90  
z                 4.04     31.80  
clarity_scale     4.00      7.00  
color_scale       5.00      6.00

In [ ]:
"""
Here we see quick several statistical definitions. 
So we can have a quick impression of the data. 
"""

You have probably noticed that the columns x, y and z have a minimum value of 0. This means that there are one or more rows (or observations) in our dataset that are supposedly representing a diamond that has lenght, width or depth of 0. Considering that we're talking about a physical object, this is impossible!

Now let's proceed to check the rows that have a value of 0 in any of the x, y or z columns. By doing this we want to check if the data we are missing can be obtained using the data that we do have.

**Check the columns with `x`, `y` and `z` with value 0 in all of them and comment what you see**

In [112]:
x_null = diamonds[diamonds['x'] == 0]
y_null = diamonds[diamonds['y'] == 0]
z_null = diamonds[diamonds['z'] == 0]
z_null

#I have used count to count the different sets of the missing values. 
# x = 8 --> Everywhere where x == 0 z is also == 0 but there is one value where y has value
# y = 7 --> Everywhere where y == 0 also x & z == 0 
# z = 20 --> sometimes there is a value for x & y where z is equal to 0

carat        cut color clarity  depth  table  price     x     y    z  \
2207    1.00    Premium     G     SI2   59.1   59.0   3142  6.55  6.48  0.0   
2314    1.01    Premium     H      I1   58.1   59.0   3167  6.66  6.60  0.0   
4791    1.10    Premium     G     SI2   63.0   59.0   3696  6.50  6.47  0.0   
5471    1.01    Premium     F     SI2   59.2   58.0   3837  6.50  6.47  0.0   
10167   1.50       Good     G      I1   64.0   61.0   4731  7.15  7.04  0.0   
11182   1.07      Ideal     F     SI2   61.6   56.0   4954  0.00  6.62  0.0   
11963   1.00  Very Good     H     VS2   63.3   53.0   5139  0.00  0.00  0.0   
13601   1.15      Ideal     G     VS2   59.2   56.0   5564  6.88  6.83  0.0   
15951   1.14       Fair     G     VS1   57.5   67.0   6381  0.00  0.00  0.0   
24394   2.18    Premium     H     SI2   59.4   61.0  12631  8.49  8.45  0.0   
24520   1.56      Ideal     G     VS2   62.2   54.0  12800  0.00  0.00  0.0   
26123   2.25    Premium     I     SI1   61.3   58.0  15397  8.52  8.42  0.0   
26243   1.20    Premium     D    VVS1   62.1   59.0  15686  0.00  0.00  0.0   
27112   2.20    Premium     H     SI1   61.2   59.0  17265  8.42  8.37  0.0   
27429   2.25    Premium     H     SI2   62.8   59.0  18034  0.00  0.00  0.0   
27503   2.02    Premium     H     VS2   62.7   53.0  18207  8.02  7.95  0.0   
27739   2.80       Good     G     SI2   63.8   58.0  18788  8.90  8.85  0.0   
49556   0.71       Good     F     SI2   64.1   60.0   2130  0.00  0.00  0.0   
49557   0.71       Good     F     SI2   64.1   60.0   2130  0.00  0.00  0.0   
51506   1.12    Premium     G      I1   60.4   59.0   2383  6.71  6.67  0.0   

       clarity_scale  color_scale  
2207               1            3  
2314               0            2  
4791               1            3  
5471               1            4  
10167              0            3  
11182              1            4  
11963              3            2  
13601              3            3  
15951              4            3  
24394              1            2  
24520              3            3  
26123              2            1  
26243              6            6  
27112              2            2  
27429              1            2  
27503              3            2  
27739              1            3  
49556              1            4  
49557              1            4  
51506              0            3

As you can see, we have 20 rows that have a value of 0 in some or all the aforementioned columns.
Most of them (12) are missing the z value, which we can obtain using the columns depth, x and y. 

20 rows with issues represent just 0.03% of our data (20 out of 53940) so it wouldn't be a big deal to remove them. Still, lets try to keep all the data we have. 

For those 12 rows, we will create a function that applies the formula given in the codebook and get the value of z. We will drop the other rows (8), since they are missing all 3 values or 2 of them.

**Create a function named `calculate_z` that applies the function in the codebook to one single row you give to the function**

In [113]:
def calculate_z (row):
    if row['z'] > 0: 
        return row['z']
    elif row['z'] == 0 and row['x'] != 0 and row['y'] != 0: 
        return row['depth']*row['x']*row['y']/200
    else: 
        return 0




**Apply it just to the rows with incorrect values**

In [116]:
diamonds['z'] = diamonds.apply(lambda row: calculate_z(row), axis=1)
diamonds.head(50)


carat        cut color clarity  depth  table  price     x     y     z  \
0    0.23      Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43   
1    0.21    Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31   
2    0.23       Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31   
3    0.29    Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63   
4    0.31       Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75   
5    0.24  Very Good     J    VVS2   62.8   57.0    336  3.94  3.96  2.48   
6    0.24  Very Good     I    VVS1   62.3   57.0    336  3.95  3.98  2.47   
7    0.26  Very Good     H     SI1   61.9   55.0    337  4.07  4.11  2.53   
8    0.22       Fair     E     VS2   65.1   61.0    337  3.87  3.78  2.49   
9    0.23  Very Good     H     VS1   59.4   61.0    338  4.00  4.05  2.39   
10   0.30       Good     J     SI1   64.0   55.0    339  4.25  4.28  2.73   
11   0.23      Ideal     J     VS1   62.8   56.0    340  3.93  3.90  2.46   
12   0.22    Premium     F     SI1   60.4   61.0    342  3.88  3.84  2.33   
13   0.31      Ideal     J     SI2   62.2   54.0    344  4.35  4.37  2.71   
14   0.20    Premium     E     SI2   60.2   62.0    345  3.79  3.75  2.27   
15   0.32    Premium     E      I1   60.9   58.0    345  4.38  4.42  2.68   
16   0.30      Ideal     I     SI2   62.0   54.0    348  4.31  4.34  2.68   
17   0.30       Good     J     SI1   63.4   54.0    351  4.23  4.29  2.70   
18   0.30       Good     J     SI1   63.8   56.0    351  4.23  4.26  2.71   
19   0.30  Very Good     J     SI1   62.7   59.0    351  4.21  4.27  2.66   
20   0.30       Good     I     SI2   63.3   56.0    351  4.26  4.30  2.71   
21   0.23  Very Good     E     VS2   63.8   55.0    352  3.85  3.92  2.48   
22   0.23  Very Good     H     VS1   61.0   57.0    353  3.94  3.96  2.41   
23   0.31  Very Good     J     SI1   59.4   62.0    353  4.39  4.43  2.62   
24   0.31  Very Good     J     SI1   58.1   62.0    353  4.44  4.47  2.59   
25   0.23  Very Good     G    VVS2   60.4   58.0    354  3.97  4.01  2.41   
26   0.24    Premium     I     VS1   62.5   57.0    355  3.97  3.94  2.47   
27   0.30  Very Good     J     VS2   62.2   57.0    357  4.28  4.30  2.67   
28   0.23  Very Good     D     VS2   60.5   61.0    357  3.96  3.97  2.40   
29   0.23  Very Good     F     VS1   60.9   57.0    357  3.96  3.99  2.42   
30   0.23  Very Good     F     VS1   60.0   57.0    402  4.00  4.03  2.41   
31   0.23  Very Good     F     VS1   59.8   57.0    402  4.04  4.06  2.42   
32   0.23  Very Good     E     VS1   60.7   59.0    402  3.97  4.01  2.42   
33   0.23  Very Good     E     VS1   59.5   58.0    402  4.01  4.06  2.40   
34   0.23  Very Good     D     VS1   61.9   58.0    402  3.92  3.96  2.44   
35   0.23       Good     F     VS1   58.2   59.0    402  4.06  4.08  2.37   
36   0.23       Good     E     VS1   64.1   59.0    402  3.83  3.85  2.46   
37   0.31       Good     H     SI1   64.0   54.0    402  4.29  4.31  2.75   
38   0.26  Very Good     D     VS2   60.8   59.0    403  4.13  4.16  2.52   
39   0.33      Ideal     I     SI2   61.8   55.0    403  4.49  4.51  2.78   
40   0.33      Ideal     I     SI2   61.2   56.0    403  4.49  4.50  2.75   
41   0.33      Ideal     J     SI1   61.1   56.0    403  4.49  4.55  2.76   
42   0.26       Good     D     VS2   65.2   56.0    403  3.99  4.02  2.61   
43   0.26       Good     D     VS1   58.4   63.0    403  4.19  4.24  2.46   
44   0.32       Good     H     SI2   63.1   56.0    403  4.34  4.37  2.75   
45   0.29    Premium     F     SI1   62.4   58.0    403  4.24  4.26  2.65   
46   0.32  Very Good     H     SI2   61.8   55.0    403  4.35  4.42  2.71   
47   0.32       Good     H     SI2   63.8   56.0    403  4.36  4.38  2.79   
48   0.25  Very Good     E     VS2   63.3   60.0    404  4.00  4.03  2.54   
49   0.29  Very Good     H     SI2   60.7   60.0    404  4.33  4.37  2.64   

    clarity_scale  color_scale  
0               1            5  
1         

If we leave the other 8 values as they are, it would negatively affect our analysis, because these are data that do not make logical sense. Therefore it is better to consider those values as NaN values, since they are probably the result of a mistake or error during process of measuring and storing these values in a dataset.

To replace them we can use the pandas .replace() method and np.NaN.

**Replace the zero values in the `z` column for a NaN**

In [ ]:
#your code here

----
# Bonus: check the new z values
Since we need to be 100% sure of our data, let's create a function that validates our z. To do so, we will use the same formula, but this time we will calculate the value of depth with the new value assigned to z.

**Create a function named `validate_z` that compares the `z`  in cells above with the one thrown by the formula and run it with the rows you changed in the cells above**

In [ ]:
#your code here

Let's check the data again with the `describe()` method.

The minimum value for x, y and z should now be a positive number, as it should be for the physical measurements of an object.

Let's finish by checking for NaN values in the data. Since we introduced them ourselves using 'replace', we will surely find some, but there may be more that are unrelated to the x, y and z columns. Checking NaNs is a fundamental part of data cleaning and it's always better to do this kind of operations before proceeding with analysis.

**Check how many NaNs do you have, comment what you would do with those values, and then do so**

In [ ]:
#your code here

# 3. Checking for outliers
Now we are going to revisit the summary table to check for outliers.

**Use the `describe` method again and comment on what you see. After that, check if you have any outliers** 

In [ ]:
#your code here

In [ ]:
#your comments here

To manage these outliers, we are going to filter our DataFrame, we're going to take all the values that have a price higher than the 75th percentile.

**Look for that quantile and filter the dataframe to clearly see the outliers. What do you think?**

In [ ]:
#your code here

Our dataset is really big and the outliers are really far apart from the rest of the values. To see this more clearly we will use a boxplot, which plots the median, 25th and 75th quartile, the maximum and minimum, as well as any outliers.

In [ ]:
#Run this code
fig, ax = plt.subplots(1,2, figsize=(10, 5))
sns.boxplot(y=diamonds.y, ax=ax[0])
sns.boxplot(y=diamonds.z, ax=ax[1])
plt.subplots_adjust(wspace=0.5)

Now we can see that all the values are within an acceptable range, but we have 2 big outliers in y and 1 in z. Now we know that our max values for y should be around 10 and the values for z should be around 6, so let's filter our dataset to find values higher than 10 in it.


In [ ]:
#your code here

Now that we have found the outlier, let's use the function we defined earlier to correct this value. First, we need to change the value to 0 (because that's how we defined the function before) and then we will apply it.

**Apply `calculate_z` for the row with the outlier**

In [ ]:
#your code here

Let's check if we actually corrected the outlier.

In [ ]:
diamonds.loc[48410]

Cool! Now let's validate our new `z`. We will check if we obtain the same value of depth using our validate function. If the formula applies, this means could approximate the real value of `z`.

**Apply `validate_z` to the row used earlier**

In [ ]:
#your code here

Now let's do the same for `y`. First, let's filter the DataFrame to find the outliers. We said that the maximum values should be around 10, so let's check what are the values above 10.

**Check the values greater than 10 in the `y` column** 

In [ ]:
#your code here

We can clearly see that the 31.8 in row 49189 is an outlier for the y value. Also, we can see that the 58.9 value for `y` in row 24067 is actually its depth, so it was a mistake when they introduced the data. Let's create a function to fix these outliers.

**Create a function named `calculate_y` to calculate `y` using `z` and `x` the same way you did above**

In [ ]:
#your code here

We will check the rows that had an outlier in `y`, to check that the values were changed.

**Check those rows (also validating with your function) and comment what you see**

Now that we have corrected or dropped all of our outliers, lets plot another box plot to double check.

In [ ]:
#Run this code
fig, ax = plt.subplots(1,2, figsize=(10, 5))
sns.boxplot(y=diamonds.y, ax=ax[0])
sns.boxplot(y=diamonds.z, ax=ax[1])
plt.subplots_adjust(wspace=0.5)

**What do you think? Are these values more reasonable?**


In [ ]:
#your thoughts here

**Once you are happy with your cleaning, save the cleaned data and continue to csv. Your new csv should be named ``diamonds_clean``**

In [ ]:
#your code here